cmd/test.go

```diff
@@ -0,0 +1,290 @@
+// Copyright © 2022-2024 Obol Labs Inc. Licensed under the terms of a Business Source License 1.1
+
+package cmd
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/spf13/cobra"
+
+	"github.com/obolnetwork/charon/app/errors"
+)
+
+type testConfig struct {
+	OutputFormat string
+	OutputFile   string
+	Quiet        bool
+	TestCases    []string
+	Timeout      time.Duration
+}
+
+func newTestCmd(cmds ...*cobra.Command) *cobra.Command {
+	root := &cobra.Command{
+		Use:   "test",
+		Short: "Test subcommands provide test suite to evaluate current cluster setup",
+		Long:  `Test subcommands provide test suite to evaluate current cluster setup. Currently there is support for peer connection tests, beacon node and validator API.`,
+	}
+
+	root.AddCommand(cmds...)
+
+	return root
+}
+
+func bindTestFlags(cmd *cobra.Command, config *testConfig) {
+	cmd.Flags().StringVar(&config.OutputFile, "output-file", "", "File path to which output can be written.")
+	cmd.Flags().StringVar(&config.OutputFormat, "output-format", "json", "File format to which output is written. Flag --output-file is required.")
+	cmd.Flags().StringSliceVar(&config.TestCases, "test-cases", nil, "List of comma separated names of tests to be exeucted.")
+	cmd.Flags().DurationVar(&config.Timeout, "timeout", 24*time.Hour, "Execution timeout for all tests.")
+	cmd.Flags().BoolVar(&config.Quiet, "quiet", false, "Do not print test results to stdout.")
+}
+
+type Duration struct {
+	time.Duration
+}
+
+func (d Duration) MarshalJSON() ([]byte, error) {
+	res, err := json.Marshal(d.String())
+	if err != nil {
+		return nil, errors.Wrap(err, "marshal json duration")
+	}
+
+	return res, nil
+}
+
+func (d *Duration) UnmarshalJSON(b []byte) error {
+	var v any
+	err := json.Unmarshal(b, &v)
+	if err != nil {
+		return errors.Wrap(err, "unmarshal json duration")
+	}
+	switch value := v.(type) {
+	case float64:
+		d.Duration = time.Duration(value)
+		return nil
+	case string:
+		d.Duration, err = time.ParseDuration(value)
+		if err != nil {
+			return errors.Wrap(err, "parse string time to duration")
+		}
+
+		return nil
+	default:
+		return errors.New("invalid duration")
+	}
+}
+
+type testVerdict string
+
+const (
+	// boolean tests
+	testVerdictNotOk testVerdict = "NotOK"
+	testVerdictOk    testVerdict = "OK"
+
+	// measurement tests
+	testVerdictGood testVerdict = "Good"
+	testVerdictAvg  testVerdict = "Avg"
+	testVerdictBad  testVerdict = "Bad"
+
+	// errored tests
+	testVerdictFail    testVerdict = "Fail"
+	testVerdictTimeout testVerdict = "Timeout"
+)
+
+type categoryScore string
+
+const (
+	categoryScoreA categoryScore = "A"
+	categoryScoreB categoryScore = "B"
+	categoryScoreC categoryScore = "C"
+)
+
+type testResult struct {
+	Verdict     testVerdict `json:"verdict"`
+	Measurement string      `json:"measurement"`
+	Suggestion  string      `json:"suggestion"`
+	Error       string      `json:"error"`
```

Comment:
Maybe we shall `omitempty` for the `error` field?

Reply:
Good one, added it